# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/media/mediacontrol/MediaController.h
# Commit: 714b2fb21e23
# Full Hash: 714b2fb21e2341e8e2bf1be1271ca0ed5e526443
# Author: alwu <alwu@mozilla.com>
# Date: 2020-06-09 09:21:34
# Regressor Bug: 1640998
# File Overlap Count: 1
# Description:
#   Bug 1640998 - part5 : set supported keys to the event source r=chunmin
#   
#   This patch will
#   - add a method `SetSupportedMediaKeys()` on `MediaControlKeysEventSource`
#   - set main controller's supported key to the event source
# ==============================================================================

diff -r a48b6554ba9c -r 714b2fb21e23 dom/media/mediacontrol/MediaController.h
--- a/dom/media/mediacontrol/MediaController.h	Mon Jun 08 22:16:46 2020 +0000
+++ b/dom/media/mediacontrol/MediaController.h	Tue Jun 09 02:23:35 2020 +0000
@@ -104,11 +104,22 @@
   // then calling any its method won't take any effect.
   void Shutdown();
 
+  // This event would be notified media controller's supported media keys
+  // change.
+  MediaEventSource<nsTArray<MediaControlKeysEvent>>&
+  SupportedKeysChangedEvent() {
+    return mSupportedKeysChangedEvent;
+  }
+
+  CopyableTArray<MediaControlKeysEvent> GetSupportedMediaKeys() const;
+
  private:
   ~MediaController();
   void HandleActualPlaybackStateChanged() override;
   void UpdateMediaControlKeysEventToContentMediaIfNeeded(
       MediaControlKeysEvent aEvent);
+  void HandleSupportedMediaSessionActionsChanged(
+      const nsTArray<MediaSessionAction>& aSupportedAction);
 
   // This would register controller to the media control service that takes a
   // responsibility to manage all active controllers.
@@ -124,6 +135,15 @@
   bool mIsRegisteredToService = false;
   bool mShutdown = false;
   bool mIsInPictureInPictureMode = false;
+
+  // We would monitor the change of media session actions and convert them to
+  // the media keys, then determine the supported media keys.
+  MediaEventListener mSupportedActionsChangedListener;
+  MediaEventProducer<nsTArray<MediaControlKeysEvent>>
+      mSupportedKeysChangedEvent;
+  // Use copyable array so that we can use the result as a parameter for the
+  // media event.
+  CopyableTArray<MediaControlKeysEvent> mSupportedKeys;
 };
 
 }  // namespace dom